src/loops.rs

```diff
@@ -283,16 +306,31 @@ fn is_iterable_array(ty: ty::Ty) -> bool {
     }
 }
 
+/// If a statement (possibly a `let` binding) consists of an expression, return it.
+fn extract_single_expr_from_stmt(block: &Block) -> Option<&Expr> {
+    match block.expr {
+        Some(_) => None,
+        None => match block.stmts[0].node {
+            StmtDecl(ref decl, _) => match decl.node {
+                DeclLocal(ref local) => match local.init {
+                    Some(ref expr) => Some(expr),
+                    None => None,
+                },
+                _ => None,
+            },
+            _ => None,
+        },
+    }
+}
+
 /// If block consists of a single expression (with or without semicolon), return it.
 fn extract_single_expr(block: &Block) -> Option<&Expr> {
-    match (&block.stmts.len(), &block.expr) {
-        (&1, &None) => match block.stmts[0].node {
-            StmtExpr(ref expr, _) |
-            StmtSemi(ref expr, _) => Some(expr),
+    match block.expr {
+        Some(ref expr) => Some(expr),
+        None => match block.stmts[0].node {
```

Comment:
What if a block has multiple statements? The comments and function name should be updated